src/framework/midi/miditypes.h

```diff
@@ -32,111 +32,87 @@
 #include "async/channel.h"
+#include "retval.h"
 #include "midievent.h"
 
 namespace mu::midi {
-using track_t = unsigned int;
-using program_t = unsigned int;
-using bank_t = unsigned int;
-using tick_t = int;
-using msec_t = uint64_t;
-using tempo_t = unsigned int;
+using track_t = int32_t;
+using program_t = int32_t;
+using bank_t = int32_t;
+using tick_t = uint32_t;
+using tempo_t = uint32_t;
 using TempoMap = std::map<tick_t, tempo_t>;
 
 using SynthName = std::string;
 using SynthMap = std::map<midi::channel_t, SynthName>;
 
 using EventType = Ms::EventType;
 using CntrType = Ms::CntrType;
-using Events = std::multimap<tick_t, Event>;
-
-struct Chunk {
-    tick_t beginTick = 0;
-    tick_t endTick = 0;
-    Events events;
-};
-using Chunks = std::map<tick_t /*begin*/, Chunk>;
+using Events = std::map<tick_t, std::vector<Event> >;
 
 struct Program {
     channel_t channel = 0;
     program_t program = 0;
     bank_t bank = 0;
-};
-using Programs = std::vector<midi::Program>;
 
-struct Track {
-    track_t num = 0;
-    std::vector<channel_t> channels;
+    bool operator==(const Program& other) const
+    {
+        return channel == other.channel
+               && program == other.program
+               && bank == other.bank;
+    }
 };
+using Programs = std::vector<midi::Program>;
 
-struct MidiData {
+struct MidiMapping {
     int division = 480;
-    TempoMap tempoMap;
-    SynthMap synthMap;
-    std::vector<Event> initEvents;  //! NOTE Set channels programs and others
-    std::vector<Track> tracks;
-    Chunks chunks;
-
-    bool isValid() const { return !tracks.empty(); }
+    TempoMap tempo;
+    SynthName synthName;
+    Programs programms;
 
-    std::set<channel_t> channels() const
+    bool isValid() const
     {
-        std::set<channel_t> cs;
-        for (const Event& e : initEvents) {
-            cs.insert(e.channel());
-        }
-        return cs;
+        return !synthName.empty() && !programms.empty() && !tempo.empty();
     }
 
-    std::vector<Event> initEventsForChannels(const std::set<channel_t>& chs) const
+    bool operator==(const MidiMapping& other) const
     {
-        std::vector<Event> evts;
-        for (const Event& e : initEvents) {
-            if (chs.find(e.channel()) != chs.end()) {
-                evts.push_back(e);
-            }
-        }
-        return evts;
+        return division == other.division
+               && tempo == other.tempo
+               && synthName == other.synthName
+               && programms == other.programms;
     }
+};
 
-    tick_t lastChunksTick() const
-    {
-        if (chunks.empty()) {
-            return 0;
-        }
-        return chunks.rbegin()->second.endTick;
-    }
+struct MidiStream {
+    tick_t lastTick = 0;
 
-    std::string dump(bool withEvents = false)
+    ValCh<std::vector<Event> > controlEventsStream;
+    async::Channel<Events, tick_t /*endTick*/> mainStream;
+    async::Channel<Events, tick_t /*endTick*/> backgroundStream;

```

Comment:
what's the difference between `mainStream` and `backgroundStream`?

Reply:
ah, it was so hard to name it. And I'm still not happy with that

mainStream - is aimed to work with the general stream of events from track
the only purpose of the "background" one is to serve for "on-click" playback


I think in the nearest future I'll rename it into "auxillaryStream" or even remove it entirely, because it was added before the idea with "EventsBuffer", it might turn out that we can achieve the same result using the single stream of events in the latest solution.